src/data/list/alist.lean

```diff
@@ -9,84 +9,84 @@ import data.list.sigma
 
 universes u v w
 open list
+variables {α : Type u} {β : α → Type v}
 
-/-- `alist α β` is a key-value map stored as a linked list.
+/-- `alist β` is a key-value map stored as a list.
```

Comment:
Why was "linked" removed? I am emphasizing the computational structure because there is a good chance of this getting used for programming

Reply:
I prefer the changed version. For me a linked list implies that we talk about pointers (i.e. concrete links between our nodes). While list makes it still clear that we talk about `list`.

Everything else has a different name, i.e. `vector`, `array`, or `buffer`

Reply:
yes, but unless you know lean internals it's not obvious that `array` and `buffer` are arrays, and `list` and `vector` are linked list. It's not trivial to keep track of all the relations, and explaining it at the type constructor seems like a good idea. Perhaps "as a `list` (i.e. a linked list)" would resolve these worries?

Reply:
I removed “linked” because, to me, `list` is what is often called a cons-list, not a linked list. Typically, linked list operations involve destructive pointer manipulation, whereas a cons-list data type is typically immutable.

Even though a cons-list may be implemented with a linked list, I don't generally see how mentioning linked list helps. The only exception that comes to mind is that someone not familiar with list types from other languages such as Haskell, ML, Coq, etc. may not immediately recognize the structure of the `inductive list` in Lean. But then I think (a) that comment should probably appear on `list` and not `alist` and (b) that person has a much larger problem than a comment on `alist` could address.

> yes, but unless you know lean internals it's not obvious that `array` and `buffer` are arrays, and `list` and `vector` are linked list.

@digama0 I don't know Lean internals, so I don't understand this statement. I know about this comment on `d_array`:

```lean
/- In the VM, d_array is implemented a persistent array. -/
structure d_array (n : nat) (α : fin n → Type u) :=
```

Are you referring to some missing comment on `list` that would say “In the VM, `list` is implemented as a linked list”? In other words, is there some special hardcoded implementation of `list`?

Reply:
Yes, I think a comment on list saying "this is implemented as a linked list" is very reasonable and it would be nice if it was there. The reason you might want to know about the data structure's internal representation is that it gives you a hint at the expected performance characteristics of various functions, like length (O(n)), nth (O(n)) and cons (O(1)) that would be different if it was an array instead.

I will grant that the lack of mutability makes it a bit less than a regular singly linked list in an imperative setting, but immutability is the default in lean so maybe users can guess that, plus I've never heard the term "cons-list" and I'm not sure whether it's more recognizable or clear than "linked list" as a shorthand description for a data structure.

As for whether there is some special hardcoded implementation for `list`: no, it's the opposite - `array` has a special implementation while `list` gets the normal implementation that all inductive types get, which is a kind of tree-with-discriminator. If you know a little about persistent data structures it's the obvious thing, but it's not a bad idea to remind people about it in particular cases, especially for really basic and foundational data structures like `list` where there is a significant possibility that it has an overridden VM representation. `string` has an overridden representation, `array` has an overridden representation, `nat` has an overridden representation, so maybe `list` does too. (It doesn't, but this is nonzero information content.)

Why would we repeat this information in derived constructions? Because they are supposed to be presenting an API, where you don't have to think about the representation. A `finmap` could have been built on `buffer` instead of `list`, and its performance characteristics would be different, even if the user-facing definitions and theorems are the same. You aren't supposed to treat a `finmap` as a `list` under normal circumstances, it has its own operations, so the underlying data structure is an implementation detail, but it is one that affects those performance characteristics. Especially after you have built up a whole hierarchy of data types, like `fintype -> finset -> multiset -> list`, it's important for some uses to know what, ultimately, this data type amounts to / is a wrapper around. I don't think we need to make a big deal about it, but I think the information should be there.

Reply:
Fair enough. Thanks for your explanation.

I think (a) that it's a matter of perspective and (b) that we're coming from different backgrounds.

I certainly agree that it's useful in some cases to know the performance characteristics of data structures and algorithms. I would never argue with that.

In my case, once I see the `list` inductive in Lean and know that Lean is strict in its evaluation, I have an idea of its performance. Calling the `list` a linked list actually distracts me from that understanding. As with @johoelzl, it makes me think of pointers and gives me the impression there might be destructive operations involve, since that is typical of linked lists. So, from my perspective, it is exactly enough to know that `alist` (or anything else) is implemented as a `list`.

One last question, @digama0, and thanks for your patience. Given this comment:

> `alist β` is a key-value map stored as a `list`.

What is, for you, a potential misinterpretation? Is it that `list` is a synonym for or wrapper around `array`? In that case, I (as a programmer for whom performance mattered) would look at the definition of `list` to see how it is defined, assuming I was not already familiar with it.

Given this comment:

> `alist α β` is a key-value map stored as a linked list.

I (again, as a programmer for whom performance mattered) would be curious about “linked list,” since I'm familiar with them. I would then have to look at the implementation of `list` and see that it doesn't appear to be the linked list I was expecting.

All that said, I'm happy to compromise on your suggestion of “as a `list` (i.e. a linked list).”

Reply:
I think that, no matter what language you come from, you would look at the definition of `list` in Lean if you are concerned about performance. So, I don't believe those assumptions hold. I also think calling it a linked list is deceptive. Nonetheless, the documentation was already there, changing it is not the intent of this PR, and we have a compromise, so I've pushed a fix in 7b4ead2078d5374066f1a8c6cd4dc3f6456b4119.